src/mono/wasm/runtime/hybrid-globalization/change-case.ts

```diff
@@ -0,0 +1,57 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+import { Module } from "../imports";
+import { setU16 } from "../memory";
+import { mono_wasm_new_external_root } from "../roots";
+import { conv_string_root } from "../strings";
+import { MonoString, MonoStringRef } from "../types";
+import { Int32Ptr } from "../types/emscripten";
+import { pass_exception_details } from "./common";
+
+export function mono_wasm_change_case_invariant(exceptionMessage: Int32Ptr, src: number, srcLength: number, dst: number, dstLength: number, toUpper: number) : void{
```

Comment:
I guess you can have faster invariant implementation in C#. Without marshaling.

Reply:
True: https://www.ibm.com/docs/en/i/7.2?topic=tables-unicode-lowercase-uppercase-conversion-mapping-table.

Reply:
The idea was interesting so I tried (implementation in C#, char by char, with a bunch of else-ifs to shift the codes where it was possible and ~30 code points that were not possible to fit to shifting rules, saved in dictionary). Still, it is slower than passign the string to JS and using there the inbuilt function.

| measurement on invariant | time pure C# | time this PR
|-:|-:|-:|
|               String, TextInfo ToLower |    12.1366ms | 1.1819ms |
|               String, TextInfo ToUpper |    11.3309ms |1.1190ms |
|            String, TextInfo ToTileCase |    26.4930ms | 23.4772ms |

Reply:
After editing the approach to use chars instead of int/double (still with the dictionary that holds exceptions from rules):

| measurement on invariant | time |
|-:|-:|
|               String, TextInfo ToLower |    17.8000ms |
|               String, TextInfo ToUpper |    17.0963ms |
|            String, TextInfo ToTileCase |    25.7467ms |

Because it does not differ much from the original approach, I am not trying to exchange dictionary to array to speed up even more because the difference would be marginal. I am leaving the original approach.

Reply:
Note that dictionaries in C# with valuetype keys can have bad performance without a custom comparer due to all the overhead involved in the default fallback comparison mode. It's especially bad in interp because we don't fully optimize the relevant code like ryujit does. So a custom comparer might make the gap close - but I suspect it's not possible to beat v8's implementation.